core/corehttp/hostname.go

```diff
@@ -151,16 +159,44 @@ func HostnameOption() ServeOption {
 					return
 				}
 
-				// Do we need to fix multicodec in PeerID represented as CIDv1?
-				if isPeerIDNamespace(ns) {
-					keyCid, err := cid.Decode(rootID)
-					if err == nil && keyCid.Type() != cid.Libp2pKey {
-						if newURL, ok := toSubdomainURL(hostname, pathPrefix+r.URL.Path, r); ok {
-							// Redirect to CID fixed inside of toSubdomainURL()
+				// Check if rootID is a valid CID
+				if rootCID, err := cid.Decode(rootID); err == nil {
+					// Do we need to redirect root CID to a canonical DNS representation?
+					dnsCID, err := toDNSPrefix(rootID, rootCID)
```

Comment:
This can convert some IPFS CIDs that are a bit too big into base36. Was that intended?

Reply:
Unsure what you mean, `toDNSPrefix` will:
1. Convert ed25519 to Base36 and return no error (because Base36 version fits in DNS label)
2. Return error if conversion to Base36 did not help and resulting CID is still too long

Reply:
Perhaps I'm confused. `toDNSPrefix` looks like 

```
	// Return as-is if things fit
	if len(rootID) <= dnsLabelMaxLength {
		return rootID, nil
	}

	// Convert to Base36 and see if that helped
	rootID, err = cid.NewCidV1(rootCID.Type(), rootCID.Hash()).StringOfBase(mbase.Base36)
	if err != nil {
		return "", err
	}
	if len(rootID) <= dnsLabelMaxLength {
		return rootID, nil
	}
```

None of this refers to Ed25519 keys or IPNS keys in general, we're just converting whatever CID we get into base36. Couldn't we have a long IPFS CID that is around the same length as the libp2p-key + identity multihash for Ed25519 keys? For example, `echo "111111111.111111111.111111" | ipfs add --inline --inline-limit=36` outputs a 65 character base32 CID. However, it'll get converted into a 62 character base 36 CID and rendered on the gateway.

My question was if this behavior is intentional

Reply:
Ah, yes, this is on purpose.
I explicitly avoided code that works only for specific codec, `toDNSPrefix` tries to fix all long CIDs.

AFAIK the switch to Base36 helps only for ED25519, but as you noted, something else could be right around the limit and I see no reason why it should not benefit from this as well.